File: python/workflows/checkout.py
"""CheckoutWorkflow — reliable execution after the agent has made a decision.

The code is intentionally ordinary: book in order, catch a failure, then undo
completed work in reverse order. Temporal makes every step durable and ensures
the compensation still runs after worker crashes or process restarts.
"""

from datetime import timedelta

from temporalio import workflow
from temporalio.common import RetryPolicy
from temporalio.exceptions import ActivityError

with workflow.unsafe.imports_passed_through():
    from activities.checkout import (
        book_activity,
        book_flight,
        book_hotel,
        cancel_activity,
        cancel_flight,
        cancel_hotel,
        finalize_checkout,
    )
    from models.types import (
        CheckoutRequest,
        CheckoutReservation,
        CheckoutResult,
        CheckoutStepRequest,
    )


CHECKOUT_RETRY = RetryPolicy(
    initial_interval=timedelta(seconds=1),
    backoff_coefficient=2.0,
    maximum_interval=timedelta(seconds=5),
    maximum_attempts=3,
    non_retryable_error_types=["HotelBookingFailed", "BookingDeclined"],
)


def _failure_message(error: ActivityError) -> str:
    return getattr(error.cause, "message", None) or "Checkout failed."


@workflow.defn
class CheckoutWorkflow:
    """Reserve an itinerary and compensate completed reservations on failure."""

    @workflow.run
    async def run(self, request: CheckoutRequest) -> CheckoutResult:
        reservations: list[CheckoutReservation] = []

        try:
            # Business ordering is explicit and easy to customize: flights
            # first, then hotels, then activities.
            for kind, booking_activity in (
                ("flight", book_flight),
                ("hotel", book_hotel),
                ("activity", book_activity),
            ):
                for item in request.items:
                    if item.kind != kind:
                        continue
                    reservation = await workflow.execute_activity(
                        booking_activity,
                        CheckoutStepRequest(account_key=request.account_key, item=item),
                        start_to_close_timeout=timedelta(seconds=30),
                        retry_policy=CHECKOUT_RETRY,
                        summary=f"Book {kind}: {item.title}",
                    )
                    reservations.append(reservation)

            booking = await workflow.execute_activity(
                finalize_checkout,
                request,
                start_to_close_timeout=timedelta(seconds=30),
                retry_policy=CHECKOUT_RETRY,
                summary="Commit completed trip",
            )
            return CheckoutResult(
                status="booked",
                message="Checkout completed and every itinerary item is booked.",
                workflow_id=workflow.info().workflow_id,
                reservations=reservations,
                booking_id=booking["booking_id"],
            )
        except ActivityError as error:
            failure = _failure_message(error)
            compensations: list[CheckoutReservation] = []

            # Saga compensation: undo only the side effects that completed,
            # in reverse order. On the demo path the hotel fails after the
            # flight succeeds, so this schedules a visible cancel_flight.
            for reservation in reversed(reservations):
                cancel_activity_for_kind = {
                    "flight": cancel_flight,
                    "hotel": cancel_hotel,
                    "activity": cancel_activity,
                }[reservation.kind]
                compensation = await workflow.execute_activity(
                    cancel_activity_for_kind,
                    reservation,
                    start_to_close_timeout=timedelta(seconds=30),
                    retry_policy=CHECKOUT_RETRY,
                    summary=f"Cancel {reservation.kind}: {reservation.title}",
                )
                compensations.append(compensation)

            cancelled = ", ".join(r.title for r in compensations) or "no prior reservations"
            return CheckoutResult(
                status="compensated",
                message=(
                    f"Checkout stopped: {failure} "
                    f"Compensation completed: cancelled {cancelled}."
                ),
                workflow_id=workflow.info().workflow_id,
                reservations=reservations,
                compensations=compensations,
                failure=failure,
            )

File: typescript/src/checkout.ts
// CheckoutWorkflow — reliable execution after the agent has made a decision.
//
// The control flow is deliberately ordinary: book in order, catch a failure,
// and undo completed work in reverse order. Temporal durably drives every step.

import {
  ActivityFailure,
  ApplicationFailure,
  proxyActivities,
  workflowInfo,
} from '@temporalio/workflow';
import type { RetryPolicy } from '@temporalio/common';
import type * as activities from './activities';
import type {
  CheckoutRequest,
  CheckoutReservation,
  CheckoutResult,
  ItineraryItem,
} from './types';

const CHECKOUT_RETRY: RetryPolicy = {
  initialInterval: '1 second',
  backoffCoefficient: 2,
  maximumInterval: '5 seconds',
  maximumAttempts: 3,
  nonRetryableErrorTypes: ['HotelBookingFailed', 'BookingDeclined'],
};

const checkoutActivities = proxyActivities<typeof activities>({
  startToCloseTimeout: '30 seconds',
  retry: CHECKOUT_RETRY,
});

function failureMessage(error: ActivityFailure): string {
  return error.cause instanceof ApplicationFailure && error.cause.message
    ? error.cause.message
    : 'Checkout failed.';
}

export async function CheckoutWorkflow(request: CheckoutRequest): Promise<CheckoutResult> {
  const reservations: CheckoutReservation[] = [];

  try {
    // Business ordering is explicit and customizable: flights, hotels, then
    // activities. Each function call becomes a durable Activity in history.
    const steps: Array<[
      ItineraryItem['kind'],
      (req: { account_key: string; item: ItineraryItem }) => Promise<CheckoutReservation>,
    ]> = [
      ['flight', checkoutActivities.book_flight],
      ['hotel', checkoutActivities.book_hotel],
      ['activity', checkoutActivities.book_activity],
    ];

    for (const [kind, book] of steps) {
      for (const item of request.items) {
        if (item.kind !== kind) continue;
        reservations.push(await book({ account_key: request.account_key, item }));
      }
    }

    const booking = await checkoutActivities.finalize_checkout(request);
    return {
      status: 'booked',
      message: 'Checkout completed and every itinerary item is booked.',
      workflow_id: workflowInfo().workflowId,
      reservations,
      compensations: [],
      booking_id: booking.booking_id as number,
    };
  } catch (error) {
    if (!(error instanceof ActivityFailure)) throw error;

    const compensations: CheckoutReservation[] = [];
    const cancelByKind = {
      flight: checkoutActivities.cancel_flight,
      hotel: checkoutActivities.cancel_hotel,
      activity: checkoutActivities.cancel_activity,
    };

    // Saga compensation: undo only completed side effects, in reverse order.
    // On the demo path the failed hotel follows a successful flight, so the
    // next event is visibly cancel_flight.
    for (const booked of [...reservations].reverse()) {
      compensations.push(await cancelByKind[booked.kind](booked));
    }

    const cancelled = compensations.map((item) => item.title).join(', ') || 'no prior reservations';
    return {
      status: 'compensated',
      message:
        `Checkout stopped: ${failureMessage(error)} ` +
        `Compensation completed: cancelled ${cancelled}.`,
      workflow_id: workflowInfo().workflowId,
      reservations,
      compensations,
      failure: failureMessage(error),
    };
  }
}
